Implement SAS Verification for crosssigning

MidhunSureshR opened the pull request.

MidhunSureshR committed:
WIP

MidhunSureshR committed:
WIP

MidhunSureshR committed:
WIP +1

bwindels commented on src/matrix/verification/SAS/stages/WaitForIncomingMessageStage.ts:
```diff
@@ -0,0 +1,73 @@
+        // todo: add timeout after 10 mins
+        return new Promise(resolve => {
+            this.track(
+                this.room._timeline.entries.subscribe({
```
Fwiw, `room._timeline` will only exist if the room is open in the UI. Either we should call `room.openTimeline()` here or we need to add some sort of callback for incoming timeline events like we added for state events for calls.

MidhunSureshR committed:
Add more stages

MidhunSureshR committed:
Free olmSas after all stages have completed

MidhunSureshR committed:
Remove comment

MidhunSureshR committed:
REFACTOR: Extract methods and functions

MidhunSureshR committed:
Emit event from DeviceMessageHandler

MidhunSureshR committed:
Remove log

MidhunSureshR committed:
Channel WIP

MidhunSureshR committed:
WIP

MidhunSureshR committed:
Upgrade olm

MidhunSureshR committed:
Implement mac and done stage

MidhunSureshR committed:
Implement send ready stage

MidhunSureshR committed:
Accept verification from device message

MidhunSureshR committed:
Throw specific error when cancelled

MidhunSureshR committed:
Implement timeout and cancel

MidhunSureshR committed:
Write a class that generates fixtures for test

MidhunSureshR committed:
Write unit tests

MidhunSureshR committed:
Refactor code

MidhunSureshR committed:
Throw error if verification was cancelled

MidhunSureshR committed:
Convert console.log to logger calls

MidhunSureshR committed:
Base stage class does not need disposable

MidhunSureshR committed:
Refactor code

1. Remove unused properties from base stage
2. Split UserData into fields
3. Write getter for channel prop

MidhunSureshR committed:
Rename class

MidhunSureshR committed:
Remove magic string

MidhunSureshR committed:
Add more logging

MidhunSureshR committed:
Fix formatting

MidhunSureshR committed:
Fix comment

MidhunSureshR committed:
Refactor SendMacStage

- Convert property to argument
- Remove unnecessary olm type
- Use Channel.getEvent
- Fix ILogItem import

MidhunSureshR committed:
Refactor SendReadyStage

- Change log identifier string

MidhunSureshR committed:
Fix string in logger

MidhunSureshR committed:
Refactor SendAcceptVerificationStage

MidhunSureshR committed:
Refactor CalculateSASStage

- Expose emoji from stage
- Await promise that resolves when emoji is matched
- Modify tests

MidhunSureshR committed:
Refactor VerifyMacStage

MidhunSureshR committed:
Refactor SASVerification class

MidhunSureshR committed:
Remove comment

MidhunSureshR committed:
Pass log last

MidhunSureshR committed:
Refactor Channel

MidhunSureshR committed:
Remove code from room vm

bwindels commented on src/matrix/DeviceMessageHandler.js:
```diff
@@ -101,6 +105,19 @@ export class DeviceMessageHandler {
+    _emitUnencryptedEvents(toDeviceEvents) {
+        const unencryptedEvents = toDeviceEvents.filter(e => e.type !== "m.room.encrypted");
+        for (const event of unencryptedEvents) {
+            this.emit("message", { unencrypted: event });
```
```suggestion
            this.emit("message", event);
```
why not just work with an optional third `DecryptionResult` argument for encrypted events? Does a bit less allocations and makes it easier to listen for messages if you don't care whether they were encrypted or not.

So below you'd have:
```
            this.emit("message", result.event, result);
```

bwindels commented on src/matrix/DeviceMessageHandler.js:
```diff
@@ -74,6 +77,7 @@ export class DeviceMessageHandler {
 
     async afterSyncCompleted(decryptionResults, deviceTracker, hsApi, log) {
+        this._emitEncryptedEvents(decryptionResults);
         if (this._callHandler) {
```
ideally we'd convert this code dealing with calls to use the event as well, but perhaps that's for a different PR...

MidhunSureshR replied:
Created https://github.com/vector-im/hydrogen-web/issues/1061

bwindels commented on src/matrix/DeviceMessageHandler.js:
```diff
@@ -101,6 +105,19 @@ export class DeviceMessageHandler {
+    }
+
+    _emitEncryptedEvents(decryptionResults) {
+        for (const result of decryptionResults) {
```
so at this stage we haven't checked that the sender key used to send the event is signed by the ed25519 device identity key as published on the homeserver. That's what the code for calls does above. So we haven't verified the identity of the sender. Given that the SAS events are not required to be encrypted, perhaps we just don't emit for now for encrypted events so we don't shoot ourselves in the foot with this once we do use it later on?

bwindels commented on src/matrix/verification/SAS/channel/types.ts:
```diff
@@ -0,0 +1,25 @@
+    Done = "m.key.verification.done",
+}
+
+export const enum CancelTypes {
```
Given the field where this is used is called `reason`, this might be a better name?
```suggestion
export const enum CancelReason {
```

bwindels commented on src/matrix/verification/SAS/channel/MockChannel.ts:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    setOurDeviceId(id: string) {
+        this.ourUserDeviceId = id;
```
why not pass this through the constructor?

MidhunSureshR committed:
Merge branch 'support-arbitrary-toast-notifications' into sas-verification

bwindels commented on src/matrix/verification/SAS/SASVerification.ts:
```diff
@@ -0,0 +1,621 @@
+    public finished: boolean = false;
+    public readonly channel: IChannel;
+    private timeout: Timeout;
+    public readonly eventEmitter: EventEmitter<SASProgressEvents> = new EventEmitter();
```
why not inherit from EventEmitter?

bwindels commented on src/matrix/verification/SAS/stages/SelectVerificationMethodStage.ts:
```diff
@@ -0,0 +1,97 @@
+            message_authentication_codes: MAC_LIST,
+            short_authentication_string: SAS_LIST,
+        };
+        await this.channel.send(VerificationEventTypes.Start, content, log);
```
Can we add a comment here saying that this should trigger the other side to send an `accept` in reply, which will move us to the next stage?

bwindels commented on src/matrix/verification/SAS/channel/Channel.ts:
```diff
@@ -0,0 +1,294 @@
+    private readonly deviceMessageHandler: DeviceMessageHandler;
+    private readonly sentMessages: Map<VerificationEventTypes, any> = new Map();
+    private readonly receivedMessages: Map<VerificationEventTypes, any> = new Map();
+    private readonly waitMap: Map<string, {resolve: any, reject: any, promise: Promise<any>}> = new Map();
```
I've introduced a `Deferred` class in my PR that would fit very well here.

bwindels commented on src/matrix/verification/SAS/channel/MockChannel.ts:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    setStartMessage(event: any): void {
```
do you need to expose this to the Stage? Since all messages are received through the channel, it could easily do this internally? Or you think it's clearer this way?

MidhunSureshR replied:
Yes, because in the case where you send the start message but also receive a start message, there's a conflict. 
This is resolved in `SelectVerificationMethodStage` which tells the channel which the "true" start message is.

bwindels replied:
aha, that makes sense, thanks!

bwindels commented on src/matrix/verification/SAS/channel/types.ts:
```diff
@@ -0,0 +1,25 @@
+export const enum VerificationEventTypes {
```
If I'm not mistaken, we usually use singular for enums?
```suggestion
export const enum VerificationEventType {
```

bwindels commented on src/matrix/verification/SAS/stages/CalculateSASStage.ts:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    private get needsToVerifyHashCommitment(): boolean {
+        if (this.channel.initiatedByUs) {
```
```suggestion
        return this.channel.initiatedByUs;
```
or even inline this?

bwindels commented on src/matrix/verification/SAS/channel/Channel.ts:
```diff
@@ -0,0 +1,294 @@
+}
+
+export interface IChannel {
+    send(eventType: string, content: any, log: ILogItem): Promise<void>;
```
```suggestion
    send(eventType: VerificationEventTypes, content: any, log: ILogItem): Promise<void>;
```

bwindels commented on src/matrix/verification/SAS/channel/MockChannel.ts:
```diff
@@ -0,0 +1,139 @@
+        const olmUtil = new this.olm.Utility();
+        const commitment = olmUtil.sha256(commitmentStr);
+        olmUtil.free();
+        acceptMessage.commitment = commitment;
```
I'm a bit confused why we modify the sent or received event here and set the commitment on it?

MidhunSureshR replied:
This is for the unit-tests, where we use hard coded messages. The commitment on them needs to change based on your public key which is generated at random by olm.

bwindels replied:
Oh, I missed that this is in the `MockChannel`, not in the real one!

MidhunSureshR committed:
Do not emit for now

MidhunSureshR committed:
Rename CancelTypes -> CancelReason

MidhunSureshR committed:
Pass device-id through options

MidhunSureshR committed:
Inherit from EventEmitter

MidhunSureshR committed:
Add explaining comment

MidhunSureshR committed:
Use deferred

MidhunSureshR committed:
VerificationEventTypes -> VerificationEventType

MidhunSureshR committed:
Inline code

MidhunSureshR committed:
Change type

bwindels reviewed (approved):
some typescript failures, but otherwise lgtm

MidhunSureshR committed:
Merge branch 'master' into sas-verification

MidhunSureshR committed:
Fix tests and code to use new data structure

MidhunSureshR committed:
Merge branch 'master' into sas-verification

MidhunSureshR committed:
Merge branch 'master' into sas-verification

MidhunSureshR committed:
Merge branch 'master' into sas-verification

bwindels committed:
Merge branch 'master' into sas-verification

bwindels commented on src/matrix/verification/CrossSigning.ts:
```diff
@@ -14,19 +14,23 @@ See the License for the specific language governing permissions and
 import type {DeviceTracker} from "../e2ee/DeviceTracker";
 import type {HomeServerApi} from "../net/HomeServerApi";
 import type {Account} from "../e2ee/Account";
+import {ILogItem} from "../../lib";
```
this wrong import crept back in again

bwindels commented on src/matrix/verification/CrossSigning.ts:
```diff
@@ -103,7 +112,27 @@ export class CrossSigning {
+            if (this.sasVerificationInProgress &&
+                (
+                    !this.sasVerificationInProgress.finished ||
+                    // If the start message is for the previous sasverification, ignore it.
```
This comment doesn't make sense to me, why can we assume here that we're dealing with a start message?

bwindels replied:
Ah because we're only processing start messages below anyway? Perhaps this can be a bit more clear.

MidhunSureshR replied:
I refactored the code in the [other PR](https://github.com/vector-im/hydrogen-web/pull/1057/commits/e198e2039607a572b9bc6cbbddd64057e46d4d1e)

bwindels committed:
fix wrong import path that crept in merge again

bwindels committed:
group imports and import types

bwindels reviewed (approved):
lgtm apart from one remaining comment
